src/FieldValidator.php

```diff
@@ -108,9 +108,25 @@ private function assertValidType(string $type, $value): bool
         return $value instanceof $type || gettype($value) === $type;
     }
 
-    private function assertValidArrayTypes(string $type, $collection): bool
+    /**
+     * @param string|array $type
+     * @param iterable $collection
+     * @return bool
+     */
+    private function assertValidArrayTypes($type, $collection): bool
     {
-        foreach ($collection as $value) {
+        foreach ($collection as $key => $value) {
+            if (is_array($type)) {
```

Comment:
I'm not sure about passing the type as an array. It's all rather implicit, and you're never actually certain about what's in that array. 

I think we've arrived at a point where a dedicated `FieldType` with some concrete implementations are in order.

A `FieldType` would be able to validate a given value. I think we'd need three or four such implementations: `ScalarType`, `ObjectType`, `GenericType` and `ArrayKeyValueType`. 

What's your opinion?

Reply:
Agree :)

Didn't wanted to invest too much upfront, was rather a quick PoC.

Well draft the PR and look at the feedback!

Reply:
For now I just replaced the `string|array` with a single `FieldType`, this is not yet your idea about `FieldType`.

Doing that would result in an even more massive refactoring and I'm not sure if I should go in that direction.